Source/Core/Core/HW/WiimoteEmu/WiimoteEmu.cpp

```diff
@@ -1083,13 +1164,150 @@ int Wiimote::CurrentExtension() const
   return m_extension->active_extension;
 }
 
-bool Wiimote::HaveExtension() const
+bool Wiimote::ExtensionLogic::ReadDeviceDetectPin()
 {
-  return m_extension->active_extension > 0;
+  return extension->active_extension ? true : false;
 }
 
-bool Wiimote::WantExtension() const
+void Wiimote::ExtensionLogic::Update()
 {
-  return m_extension->switch_extension != 0;
+  // Update controller data from user input:
+  // Write data to addr 0x00 of extension register
+  extension->GetState(reg_data.controller_data);
+}
+
+// TODO: move this to Common if it doesn't exist already?
+template <typename T>
+void SetBit(T& value, u32 bit_number, bool bit_value)
+{
+  bit_value ? value |= (1 << bit_number) : value &= ~(1 << bit_number);
+}
+
+void Wiimote::MotionPlusLogic::Update()
+{
+  if (!IsActive())
+  {
+    return;
+  }
+
+  // TODO: clean up this hackery:
+  // the value seems to increase based on time starting after the first read of 0x00
+  if (IsActive() && times_updated_since_activation < 0xff)
+  {
+    ++times_updated_since_activation;
+
+    // TODO: wtf is this value actually..
+    if (times_updated_since_activation == 9)
+      reg_data.initialization_status = 0x4;
+    else if (times_updated_since_activation == 10)
+      reg_data.initialization_status = 0x8;
+    else if (times_updated_since_activation == 18)
+      reg_data.initialization_status = 0xc;
+    else if (times_updated_since_activation == 53)
+      reg_data.initialization_status = 0xe;
+  }
+
+  auto& mplus_data = *reinterpret_cast<wm_motionplus_data*>(reg_data.controller_data);
+  auto& data = reg_data.controller_data;
+
+  // TODO: make sure a motion plus report is sent first after init
+
+  // On real mplus:
+  // For some reason the first read seems to have garbage data
+  // is_mp_data and extension_connected are set, but the data is junk
+  // it does seem to have some sort of pattern though, byte 5 is always 2
+  // something like: d5, b0, 4e, 6e, fc, 2
```

Comment:
if it's actually reads from uninitialized memory, replace the values with 0xA5A5A5 or something.

Reply:
Sorry. I mean the REAL hardware returns weird data for the first frame. Consistently.